File: Cpp/746.cpp
class Solution {
public:
    int f(int i, vector<int>& cost, vector<int>& dp){
        if(i == 0 || i == 1)return cost[i];
        if(dp[i] != -1)return dp[i];
        return dp[i] = (cost[i] + min(f(i-1, cost, dp), f(i-2, cost, dp)));
    }
    int minCostClimbingStairs(vector<int>& cost) {
        vector<int> dp(cost.size(), -1);
        return min(f(cost.size()-1, cost, dp), f(cost.size()-2, cost, dp));
    }
};

File: Python/746.py
class Solution:
    def f(self, i, cost, dp):
        if i >= len(cost):
            return 0
        if i in dp:
            return dp[i]
        dp[i] = min(cost[i] + self.f(i+1,cost, dp), cost[i] + self.f(i+2,cost, dp))
        return dp[i]

    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = {}
        return min(self.f(0,cost, dp), self.f(1,cost, dp))
